msg_encoding_test.go

```diff
@@ -0,0 +1,156 @@
+	require.NoError(b, err)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if got, err := encoder.Decode(data); err != nil {
+				require.NoError(b, err)
+				require.Equal(b, msg, got)
+			}
+		}
+	})
+}
+
+func BenchmarkZstdEncoding(b *testing.B) {
+	rng := rand.New(rand.NewSource(seed))
+	encoder, err := newZstdGMessageEncoding()
+	require.NoError(b, err)
+	msg := generateRandomGMessage(b, rng)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if _, err := encoder.Encode(msg); err != nil {
+				require.NoError(b, err)
+			}
+		}
+	})
+}
+
+func BenchmarkZstdDecoding(b *testing.B) {
+	rng := rand.New(rand.NewSource(seed))
+	encoder, err := newZstdGMessageEncoding()
+	require.NoError(b, err)
+	msg := generateRandomGMessage(b, rng)
+	data, err := encoder.Encode(msg)
+	require.NoError(b, err)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if got, err := encoder.Decode(data); err != nil {
+				require.NoError(b, err)
+				require.Equal(b, msg, got)
+			}
+		}
+	})
+}
+
+func generateRandomGMessage(b *testing.B, rng *rand.Rand) *gpbft.GMessage {
+	return &gpbft.GMessage{
+		Sender:        gpbft.ActorID(rng.Uint64()),
+		Vote:          generateRandomPayload(b, rng),
+		Signature:     generateRandomBytes(b, rng, 96),
+		Ticket:        generateRandomBytes(b, rng, 96),
+		Justification: generateRandomJustification(b, rng),
+	}
+}
+
+func generateRandomJustification(b *testing.B, rng *rand.Rand) *gpbft.Justification {
+	return &gpbft.Justification{
+		Vote:      generateRandomPayload(b, rng),
+		Signers:   generateRandomBitfield(rng),
+		Signature: generateRandomBytes(b, rng, 96),
+	}
+}
+
+func generateRandomBytes(b *testing.B, rng *rand.Rand, n int) []byte {
+	buf := make([]byte, n)
+	_, err := rng.Read(buf)
+	require.NoError(b, err)
+	return buf
+}
+
+func generateRandomPayload(b *testing.B, rng *rand.Rand) gpbft.Payload {
+	return gpbft.Payload{
+		Instance: rng.Uint64(),
+		Round:    rng.Uint64(),
+		Phase:    gpbft.Phase(rng.Intn(3)),
+		Value:    generateRandomECChain(b, rng, 32),
+		SupplementalData: gpbft.SupplementalData{
+			PowerTable: generateRandomCID(b, rng),
+		},
+	}
+}
+
+func generateRandomBitfield(rng *rand.Rand) bitfield.BitField {
+	return bitfield.NewFromSet([]uint64{rng.Uint64(), rng.Uint64(), rng.Uint64()})
+}
+
+func generateRandomECChain(b *testing.B, rng *rand.Rand, length int) gpbft.ECChain {
+	chain := make(gpbft.ECChain, length)
+	for i := 0; i < length; i++ {
+		chain[i] = generateRandomTipSet(b, rng, int64(i))
+	}
+	return chain
+}
+
+func generateRandomTipSet(b *testing.B, rng *rand.Rand, epoch int64) gpbft.TipSet {
+	return gpbft.TipSet{
+		Epoch:      epoch,
+		Key:        generateRandomTipSetKey(b, rng),
+		PowerTable: generateRandomCID(b, rng),
+	}
+}
+
+func generateRandomTipSetKey(b *testing.B, rng *rand.Rand) gpbft.TipSetKey {
+	key := make([]byte, 20*38)
```

Comment:
20? average tipset aims for 5 blocks

Reply:
The number 20 comes from the max tipset key length constant in gpbft. I have randomised all fixed values in the generators to pick a size/length at random within configured max.